# Structured Programming

## Lab 8. Subprograms and Recursion

### Learning Objectives

By the end of this worksheet, you will be able to:

* Differentiate between functions and procedures.
* Identify the basic structure of subprograms.
* Create functions and procedures in Python.
* Understand the concept of recursion.
* Implement simple recursive functions.



# Part 1. Reviewing Functions

Observe the following code and execute it.



In [ ]:
def calculate_final_grade(exam1, exam2, project):
  return (exam1 + exam2 + project) / 3

final_grade = calculate_final_grade(85, 92, 88)
print(final_grade)

88.33333333333333




### Reflection Questions

1. What value does the function return? 88.33333333333333

---

2. Where is the result stored? final_grade

---

3. What would happen if the `return` statement were removed? None

---



# Part 2. What Is a Procedure?

In structured programming, there are two main types of subprograms:

## Function

A function performs a task and returns a value.

Example:



In [ ]:
def calculate_area(base, height):
    return base * height

area = calculate_area(5, 4)

print(area)

20




## Procedure

A procedure performs an action but does not return a value.

Example:



In [ ]:
def display_menu():
    print("1. Add")
    print("2. Subtract")
    print("3. Exit")

display_menu()

1. Add
2. Subtract
3. Exit




### Activity 1

Execute the following code.



In [ ]:
def greet(name):
    print("Hello", name)

result = greet("Anna")

print(result)



### Questions

1. What appears first on the screen? Hello Anna

---

2. What is stored in the variable `result`? None

---

3. Does the subprogram return a value? No, it is a procedure that only performs the action of printing.

---



# Part 3. Identifying Functions and Procedures

Analyze each subprogram.



```python
def get_name():
    return input("Name: ")

def print_report():
    print("Report generated")

def calculate_average(a, b, c):
    return (a + b + c) / 3

def display_message():
    print("Welcome")
```

Complete the table.

| Subprogram        | Function or Procedure |
| ----------------- | --------------------- |
| get_name          | Function              |
| print_report      | Procedure             |
| calculate_average | Function              |
| display_message   | Procedure             |




# Part 4. Creating Procedures

## Guided Exercise

Complete the following procedure.




In [ ]:
def display_header():
    print("University")
    print("Structured Programming")
    print("Subprogram Practice")
    pass

display_header()

University
Structured Programming
Subprogram Practice



Expected output:

```text
University
Structured Programming
Subprogram Practice
```

---

## Challenge

Create a procedure called `display_table_of_5()` that prints the multiplication table of 5.

Expected output:

```text
5 x 1 = 5
5 x 2 = 10
...
5 x 10 = 50
```

In [ ]:
def display_table_of_5():
    for i in range(1, 11):
        print(f"5 x {i} = {5 * i}")

display_table_of_5()

In [ ]:
def square(number):
    return number * number



print(square(4))




Expected output:

```text
16
```



## Challenge

Create a function called `is_even(number)`.

The function must return:

* True if the number is even.
* False if the number is odd.

Test your function:

```python
print(is_even(8))
print(is_even(7))
```



In [ ]:
def is_even(number):
    if number % 2 == 0:
        return True
    else:
        return False

print(is_even(8))
print(is_even(7))



# Part 6. Introduction to Recursion

A recursive function is a function that calls itself.

Observe the following example.



In [ ]:
def countdown(n):

    print(n)

    if n > 1:
        countdown(n - 1)

countdown(5)



Execute the program.

### Questions

1. Which numbers are displayed? 5, 4, 3, 2, and 1

---

2. What happens when `n` becomes 1? The program prints 1, but because the condition n > 1 is false, it does not execute the recursive call again.

---

3. Why does the program stop? It stops because the if n > 1 block acts as a stopping condition, preventing the function from calling itself infinitely.

---



# Part 7. Understanding the Base Case

Every recursive function needs a condition that stops the recursion.

Observe the following example.



In [ ]:
def say_hello(n):

    if n == 0:
        return

    print("Hello")
    say_hello(n - 1)

say_hello(4)

Hello
Hello
Hello
Hello


### Questions

1. How many times is "Hello" printed? is printed 4 times.

---

2. What is the base case? if n == 0

---

3. What would happen if the `if` statement were removed? RecursionError

---


# Part 8. Building a Recursive Function

Complete the following function.



In [ ]:
def count_to_one(n):

    print(n)

    if n > 1:
        count_to_one(n - 1)

count_to_one(10)

Observe carefully how the function calls itself.

### Reflection

What is the role of the statement:

```python
count_to_one(n - 1)
```

*Answer: It is the recursive step. It calls the same function but decreases the argument by 1 each time, moving the function closer to the stopping condition.*
---


# Part 9. First Recursive Mathematical Example

The factorial of a number is defined as:

```text
4! = 4 × 3 × 2 × 1
```

Observe the following code.

```python
def factorial(n):

    if n == 1:
        return 1

    return n * factorial(n - 1)

print(factorial(4))
```

### Activity

Complete the table.

| Function Call | Waiting For |
| ------------- | ----------- |
| factorial(4)  |4 * factorial(3)|
| factorial(3)  |3 * factorial(2)|
| factorial(2)  |2 * factorial(1)|
| factorial(1)  |1            |

---

### Reflection

Why does the function eventually stop?
It stops because it hits the base case if n == 1:, which returns 1 directly instead of executing another recursive call.
---

---

---

# Part 10. Final Challenge

Create a program that includes:

## Procedures

* display_menu()
* say_goodbye()

## Functions

* calculate_area()
* calculate_average()

## Recursive Function

* countdown()

The menu should allow the user to select an option.

```text
1. Calculate Area
2. Calculate Average
3. Countdown
4. Exit
```

### Requirements

* Use at least two procedures.
* Use at least two functions that return values.
* Use one recursive function.
* Organize the program into subprograms.


1. What is the difference between a function and a procedure?
A function performs a specific task and returns a calculated value. A procedure performs an action (like printing to the screen) but does not return a value.

2. What is recursion?
Recursion is a programming technique where a function solves a problem by calling itself.

3. Why does every recursive function need a base case?
A condition that stops the recursion is necessary to prevent the function from calling itself infinitely, which would lead to a program crash.

4. Which topic was most difficult for you?
Defining the logic for how recursive functions hold variables in memory while waiting for the base case is often the most complex topic to grasp initially.

In [ ]:
 Code of the Part 10
def display_menu():
    print("\n--- Menu ---")
    print("1. Calculate Area")
    print("2. Calculate Average")
    print("3. Countdown")
    print("4. Exit")

def say_goodbye():
    print("Thank you for using the program. Goodbye!")

def calculate_area(base, height):
    return base * height

def calculate_average(n1, n2, n3):
    return (n1 + n2 + n3) / 3

def countdown(n):
    print(n)
    if n > 1:
        countdown(n - 1)

def main():
    while True:
        display_menu()
        choice = input("Select an option (1-4): ")

        if choice == '1':
            b = float(input("Enter base: "))
            h = float(input("Enter height: "))
            print(f"Area: {calculate_area(b, h)}")

        elif choice == '2':
            num1 = float(input("Enter number 1: "))
            num2 = float(input("Enter number 2: "))
            num3 = float(input("Enter number 3: "))
            print(f"Average: {calculate_average(num1, num2, num3)}")

        elif choice == '3':
            start = int(input("Enter starting number for countdown: "))
            countdown(start)

        elif choice == '4':
            say_goodbye()
            break

        else:
            print("Invalid option. Please try again.")


main()

# References

Downey, A. (2024). *Think Python: How to Think Like a Computer Scientist*. Green Tea Press.

Python Software Foundation. *Python Documentation*. Available at: https://docs.python.org/

Gaddis, T. (2023). *Starting Out with Python* (6th Edition). Pearson.


# Submission Instructions

To receive credit for this activity, complete the following steps:

1. Make a copy of the notebook and save it in your personal repository.

2. Complete all sections of the worksheet:

   * Run every example provided.
   * Answer all reflection questions.
   * Complete all guided exercises.
   * Complete all challenge activities.

3. Verify that all code cells execute correctly before submitting.

4. Once the notebook is finished, mark the assignment as **Completed** in Google Classroom.



# Additional Practice

You are encouraged to improve your solutions by adding:

* Input validation
* Error handling
* Additional menu options
* More descriptive output messages

These improvements are optional but highly recommended as a way to strengthen your programming skills and gain additional practice with structured programming concepts.
